### DISCUSSION TITLE
Deno.run executes some programs immediately but others

### DISCUSSION THREAD

**Discussion by sant123** [denoland/deno#11723] [Q&A] [answered]
Hi, please take a look at this code:

``` ts
// mod.ts
const p1 = Deno.run({
  cmd: ['echo', 'deno'],
});

const p2 = Deno.run({
  cmd: ['deno', 'run', 'child.ts'],
});

// p2.status();
```
``` ts
// child.ts
console.log('Hi from child!');
```

If running this code with `deno run --allow-run mod.ts` it will print in stdout `deno`.

Now comment out `p2.status();` and run it again:

``` ts
// mod.ts
const p1 = Deno.run({
  cmd: ['echo', 'deno'],
});

const p2 = Deno.run({
  cmd: ['deno', 'run', 'child.ts'],
});

p2.status();
```

Now prints both `deno` and `Hi from child!`.

My question is, is this expected behavior? If it is please some one can explain me why this happens?

Thank you so much.

Deno v1.13.0

**Comment by littledivy** [comment]
`Deno.run` won't wait for the process to complete and will exit. The first process was too short lived that you could see the output before Deno exited. When you add `p.status()`, it returns a new Promise that will resolve when the process returns a status code.

**Reply by sant123** [reply]
So in theory the second process just runs and exit? I tried to write a file in the second process but didn't work either. It's like the second process never gets executed.

**Reply by nayeemrmn** [reply]
> `Deno.run` won't wait for the process to complete and will exit. The first process was too short lived that you could see the output before Deno exited. When you add `p.status()`, it returns a new Promise that will resolve when the process returns a status code.

@littledivy This seems buggy, `Deno.run()` should dispatch _some_ kind of async op that leaves the process alive right? But I wonder why this doesn't happen with `Deno.spawnChild()`...

**Reply by lucacasonato** [reply]
@nayeemrmn With `Deno.run` the process will only wait if you call `p.status()`. With `Deno.spawnChild` this happens automatically, and the promise is available on `Child#status`

**Reply by nayeemrmn** [reply]
> @nayeemrmn With `Deno.run` the process will only wait if you call `p.status()`. With `Deno.spawnChild` this happens automatically, and the promise is available on `Child#status`

Ah, we prevent it from closing stdin in that case with `stdin.take()`. So shouldn't we do the exact same thing with `Deno.run()`? And manually close it in `process.status()` to preserve the current anti-deadlock behaviour (which `Deno.spawnChild()` doesn't have).

**Comment by sant123** [comment, marked as answer]
This is solved by `Deno.spawn`:

``` ts
const p1 = await Deno.spawn("echo", {
  args: ["deno"],
  stdout: "inherit",
});

const p2 = await Deno.spawn("deno", {
  args: ["run", "child.ts"],
  stdout: "inherit",
});

```

![image](https://user-images.githubusercontent.com/7959437/166407144-c0c0c352-11cb-401d-ad42-185a7d3efa86.png)
